internal/privatestate/data.go

```diff
@@ -0,0 +1,200 @@
+
+	for k, v := range d.Framework {
+		mergedMap[k] = v
+	}
+
+	for k, v := range d.Provider {
+		mergedMap[k] = v
+	}
+
+	bytes, err := json.Marshal(mergedMap)
+	if err != nil {
+		diags.AddError(
+			"Error Encoding Private State",
+			fmt.Sprintf("An error was encountered when encoding private state: %s.\n\n"+
+				"This is always a problem with Terraform or terraform-plugin-framework. Please report this to the provider developer.", err),
+		)
+
+		return nil, diags
+	}
+
+	return bytes, diags
+}
+
+// NewData creates a new Data based on the given slice of bytes.
+// It must be a JSON encoded slice of bytes, that is map[string][]byte.
+func NewData(ctx context.Context, data []byte) (*Data, diag.Diagnostics) {
+	var (
+		dataMap map[string][]byte
+		diags   diag.Diagnostics
+	)
+
+	if len(data) == 0 {
+		return nil, nil
+	}
+
+	err := json.Unmarshal(data, &dataMap)
+	if err != nil {
+		diags.AddError(
+			"Error Decoding Private State",
+			fmt.Sprintf("An error was encountered when decoding private state: %s.\n\n"+
+				"This is always a problem with Terraform or terraform-plugin-framework. Please report this to the provider developer.", err),
+		)
+
+		return nil, diags
+	}
+
+	output := Data{
+		Framework: make(map[string][]byte),
+		Provider:  make(map[string][]byte),
+	}
+
+	for k, v := range dataMap {
+		if isInvalidProviderDataKey(ctx, k) {
+			output.Framework[k] = v
+			continue
+		}
+
+		output.Provider[k] = v
+	}
+
+	return &output, diags
+}
+
+// ProviderData contains private state data for provider usage.
+type ProviderData map[string][]byte
+
+// GetKey returns the private state data associated with the given key.
+//
+// If the key is reserved for framework usage, an error diagnostic
+// is returned. If the key is valid, but private state data is not found,
+// nil is returned.
+//
+// The naming of keys only matters in context of a single resource,
+// however care should be taken that any historical keys are not reused
+// without accounting for older resource instances that may still have
+// older data at the key.
+func (d *ProviderData) GetKey(ctx context.Context, key string) ([]byte, diag.Diagnostics) {
+	if d == nil {
+		return nil, nil
+	}
+
+	diags := ValidateProviderDataKey(ctx, key)
+
+	if diags.HasError() {
+		return nil, diags
+	}
+
+	value, ok := (*d)[key]
+	if !ok {
+		return nil, nil
+	}
+
+	return value, nil
+}
+
+// SetKey sets the private state data at the given key.
+//
+// If the key is reserved for framework usage, an error diagnostic
+// is returned. The data must be valid JSON and UTF-8 safe or an error
+// diagnostic is returned.
+//
+// The naming of keys only matters in context of a single resource,
+// however care should be taken that any historical keys are not reused
+// without accounting for older resource instances that may still have
+// older data at the key.
+func (d *ProviderData) SetKey(ctx context.Context, key string, value []byte) diag.Diagnostics {
```

Comment:
Does `SetKey` need to be declared on a pointer receiver given that `ProviderData` is a `map[string][]byte`?

Reply:
My best recommendation would be to write unit testing that verifies the expected map after a key has been set, e.g. setting a new key should verify that the new key was added or updating an existing key should verify that the value was updated. Both you and future code travelers can understand and verify the expected behaviors via that unit testing. 👍